### DISCUSSION TITLE
Transactions involving multiple datasources in Keycloak

### DISCUSSION THREAD

**Discussion by yrodiere** [keycloak/keycloak#30929] [Archive]
Hey all, @shawkins,

Some time ago, [a breaking change happened in Quarkus][1], causing exceptions at runtime, and this led to the [discovery of unsafe paths in Keycloak][2]. When going through these paths, Keycloak could involve multiple non-XA datasources in the same transaction, which could lead to a transaction being committed in some datasources, but rolled back in others.

The breaking change (causing the runtime exceptions) was reverted in [Quarkus 3.8.5, which Keycloak upgraded to][3], so it no longer affects Keycloak, and [a flag was added to Quarkus][4] so that any unsafe paths can be reported as warnings instead of triggering exceptions.

However, as far as I know, the unsafe paths remain. We [discussed this at length][5] with @shawkins, and from what I understand:

1. [XA is not an option at the moment, for lack of stable storage for the transaction log][6]
2. Some of the unsafe paths are actually read-only, so safe in practice since they won't commit any change.
Having [a way to mark transactions as read-only][7] -- and thus make Narayana understand that it's safe to involve multiple non-XA resources  -- would help .
3. Some of the [unsafe paths involve writes][8] (like granting roles to all users) and thus might really be unsafe.

I wanted us to continue our conversation on this topic, especially with @shawkins:

1. Have you learned more about the problem, and do you currently have any solutions?
2. Do you still think changes would be necessary in Quarkus to address this? Would those changes include read-only transactions, something else?

Thank you,

Yoann (Hibernate, Quarkus)

CC: @gsmet, @Sanne,  @mmusgrov

[1]: https://github.com/quarkusio/quarkus/issues/39283
[2]: https://github.com/keycloak/keycloak/issues/28725
[3]: https://github.com/keycloak/keycloak/pull/30173
[4]: https://github.com/quarkusio/quarkus/pull/40365
[5]: https://github.com/quarkusio/quarkus/pull/40365#discussion_r1586209813
[6]: https://github.com/quarkusio/quarkus/pull/40365#discussion_r1589227693
[7]: https://github.com/quarkusio/quarkus/issues/6414#issuecomment-2096008869
[8]: https://github.com/quarkusio/quarkus/pull/40365#discussion_r1587571358

**Comment by shawkins** [comment]
Thank you @yrodiere for following up on this.

> XA is not an option at the moment

That is correct we lack built-in support with operator based usage. We also officially support at least 1 database that lacks XA support.

> Some of the unsafe paths are actually read-only, so safe in practice since they won't commit any change.

That is correct - at least from a built-in perspective. Almost all of our existing code paths are either writing to a single datasource, or are read-only across all sources. There is a concern that custom integration could do anything under the transaction though.

> Some of the https://github.com/quarkusio/quarkus/pull/40365#discussion_r1587571358 (like granting roles to all users) and thus might really be unsafe.

This was where we had wanted heuristic mixed exceptions to help expose a situation that did not commit successfully. @keycloak/core should weigh in here as well as to the transactional expectations - and how we expect users choosing to run in non-xa mode (possibly because of non-xa datasource(s)) to recover.

> Have you learned more about the problem, and do you currently have any solutions?

No, we are just going to rely on the workaround introduced to support the legacy behavior.

> Do you still think changes would be necessary in Quarkus to address this? Would those changes include read-only transactions, something else?

Foremost we need to avoid hard errors, and to make users aware if things may be in an inconsistent state - we have a workaround for the former, and a log message on the latter. 

I'm open to additional changes, but it's hard to envision exactly what that would be. We have our own transaction start / commit as part of the request / response cycle outside of more typical quarkus handling. At the time we're starting the transaction we don't currently have captured what type of path we are dealing with (ready-only, write to one and what particular source, or write to multiple). Ignoring the possiblity of custom logic introducing additional writes, could identify all built-in logic that falls into the write to multiple category and make targeted logic changes there.

**Reply by yrodiere** [reply]
> I'm open to additional changes, but it's hard to envision exactly what that would be

I think the current problems we want to tackle are:

1. The false positives when warning users: if all the user is doing is read from a datasource, it doesn't matter if there are other non-XA datasources.
2. The possibility of actually unsafe paths in Keycloak -- be it in the core, or with extensions: if some path is trying to write to multiple datasources, there is an actual risk of data loss or corruption.

I am not sure we can completely solve these problems, but we might at least be able to reduce them.

---

Problem number 1 is definitely something Quarkus and (more importantly) Narayana/Agroal could help with.

The Narayana team has been [discussing solutions on the Jakarta Transactions side](https://github.com/jakartaee/transactions/issues/58#issuecomment-2098094017), but before this can reach Keycloack we'll also need to flesh out some API: how do you declare a datasource as read-only? A flag on transaction start (for all datasources) or in datasource configuration (for all uses of that datasource) are both options that could make sense for some use cases, but probably not for Keycloack, as their "scope" is too wide. Something on connection retrieval (`connection.setReadOnly()`?) may make more sense here? Or maybe something on datasource injection (`@ReadOnlyTransactions` annotation on the injection point?) would be enough?

Regardless of the solution, my point is: if you could provide examples of code paths that could take advantage of this read-only datasource feature, we could draft an API, infer what is needed in Agroal, then Narayana, and get the ball rolling.

---
Problem number 2 is much tougher, because there is no obvious solution. I think it will require more investigation and assessment on your side before we can even talk about helping in Quarkus: we'd need to know what the problematic use cases are exactly.

Depending on the exact use cases, the solution could simply be to document that such use cases are supported only with XA, and plan the switch to "safe by default" (errors instead of warnings) in a future version of Keycloak -- while exposing a configuration flag for users who have no other option than the unsafe behavior.
As you can imagine, such a switch would require that we are absolutely sure there are no false positives in errors being reported -- which makes solving problem 1 all the more important.

If the use cases are critical and cannot suffer such high requirements, one might want to try to redesign it: XA transactions are the safest and most straightforward solution, but there are other avenues worth exploring, such as compensating transactions. Those could be a fallback solution -- again, if the use cases are really worth the work this will require.

**Reply by shawkins** [reply]
> Regardless of the solution, my point is: if you could provide examples of code paths that could take advantage of this read-only datasource feature, we could draft an API, infer what is needed in Agroal, then Narayana, and get the ball rolling.

We start a transaction at the beginning of the request processing cycle. Then there is handling when the response is complete to commit or rollback. See:
https://github.com/keycloak/keycloak/blob/e79d10e71e286a7dc6787b1da98388fd0d67b16a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/resteasy/TransactionalSessionHandler.java#L36 - with the backing implementation / abstraction over JTA, JPA, etc. - https://github.com/keycloak/keycloak/blob/e79d10e71e286a7dc6787b1da98388fd0d67b16a/services/src/main/java/org/keycloak/services/DefaultKeycloakTransactionManager.java#L90

So currently we are not using any transaction related annotations, nor quarkus level transaction management.

It seems plausible that for the handful of built-in paths that could update multiple datasources that we could annotate those endpoints and pass that down somehow. Eveything else (nominally a single update, or all read-only case) we'd want to be free of warnings / errors.

**Reply by yrodiere** [reply]
> So currently we are not using any transaction related annotations, nor quarkus level transaction management.

Okay, I think this rules out marking a *transaction* as read-only. There might be other solutions though, see below.

> It seems plausible that for the handful of built-in paths that could update multiple datasources that we could annotate those endpoints and pass that down somehow

What would be the semantics of that annotation? IIUC we're talking of cases where datasources are read/write, so I'm not sure what's the point here... ?

> Eveything else (nominally a single update, or all read-only case) we'd want to be free of warnings / errors.

Right, well in order to not get any warning / error, Narayana needs to be able to tell for sure that at most one non-XA datasoure is written to.
Seems to me that would involve marking either datasources, or just connections to datasources, as read-only?
It depends how these datasources and connections are manipulated in Keycloack, though.

**Reply by yrodiere** [reply]
> > What would be the semantics of that annotation? IIUC we're talking of cases where datasources are read/write, so I'm not sure what's the point here... ?
> 
> The semantics from Keycloak's perspective would be - with the annotation there would be a warning or hard error, depending upon the keycloak config, when more than 1 non-XA datasource is involved in the transaction. Without the annotation, no warning or error is expected.

That seems backwards, as this would give an "unsafe-by-default" sort of behavior. Better annotate endpoints that are known to be safe (e.g. with `@ReadOnly` or `@AssumeAtMostOneWriteableDataSource` or whatever), and let warnings pop up anywhere there is no annotation. That way, if a new endpoint is added, at least you'll get warned about potential misbehavior.

Two remarks though -- regardless of whether we go with safe-by-default or unsafe-by-default:

1. This requires marking transactions as read-only (or as "at-most-one-writeable-datasource"), because the warnings/errors are issued by Narayana. It will require some information passing between endpoint annotations and transaction management. So if Quarkus is to do pass that information, Quarkus transaction management will need to be involved.
2. Marking an endpoint (and thus any nested transaction) as read-only may be doable, but marking it as "at-most-one-writeable-datasource" probably isn't, because Narayana needs to order datasource commits in a certain order (writeable datasource last) to make it safe (ish). I _think_ just suppressing warnings in that case would still leave you with a risk of partial commits (e.g. if writes get committed, but then a read-only datasource fails to commit due to concurrent changes).

So, at this point is seems only two avenues are worth pursuing:

1. In some cases, mark a whole endpoint as fully readonly, e.g. annotate it with a `@ReadOnly` annotation to tell Agroal/Narayana "any datasource accessed in this scope should be accessed in read-only mode". That would allow Agroal/Narayana to cooperate to make things safe (set connections as read-only) and suppress warnings.
2. In other cases, mark as read-only a given datasource (configuration properties), or datasource injection point (`@ReadOnly` on the injection point), or connection (manual call to `setReadOnly`? Only makes sense if you manipulate connections manually, or through Hibernate ORM). That would allow Narayaba to detect when at most one datasource is writeable, order commits accordingly, and suppress warnings.

**Reply by shawkins** [reply]
> That seems backwards, as this would give an "unsafe-by-default" sort of behavior.

I'm relaying this from Keycloaks perspective - what is the least effort, and what matches our current assumptions. I don't mean to imply that this annotation would be something that Quarkus or Narayana would directly know about - since we only have 1 or two resource calls that are affected (without considering custom logic) so it is much more straight-forward for us to annotate those rather than everything else. The handling of that annotation would then presumably be in the logic I linked to above in our session handler. However if this seems too short-sighted, I'm fine with being safe by default.

We definitely require number 2. The typical scenario there is an update to a federated resource, such as a user store. That update will eventually be made into just a single datasource, but that may be difficult to determine which datasource at the level at which we are starting the transaction. I believe many of these calls may require reads against the primary keycloak datasource as well.

**Reply by yrodiere** [reply]
> The handling of that annotation would then presumably be in the logic I linked to above in our session handler. However if this seems too short-sighted, I'm fine with being safe by default.

Sorry if I was dismissive; it's a decent workaround, it's just that we appear to approach the problem differently. I'm focusing on having warnings report actual problems nobody knows about, you're focusing on removing pesky, unnecessary warnings -- which is equally important and actually necessary to get to my goal.

Implementation-wise, your solution will require Narayana to display warnings or not depending on the "endpoint context". Right now all Narayana offers is a global flag spanning all requests/threads. I suspect the change will not be straightforward, and may not be high-priority considering it's a workaround, but you can certainly try to file an issue to Narayana.

> We definitely require number 2. The typical scenario there is an update to a federated resource, such as a user store. That update will eventually be made into just a single datasource, but that may be difficult to determine which datasource at the level at which we are starting the transaction. 

I think you don't need to know whether a datasource will be read-write or read-only at the time you start the transaction in the transaction manager, but a bit later, when the datasource joins the transaction (i.e. when you open a connection). Would that be easier?

Note another constraint is that you can't upgrade from read-only to read-write after joining the transaction, so you'll have to look for that. It might be okay to open a second , concurrent connection to the same datasource with read-write access, but then you might run in classic connection pool issues when acquiring multiple connections (deadlocks).

**Comment by shawkins** [comment]
> What would be the semantics of that annotation? IIUC we're talking of cases where datasources are read/write, so I'm not sure what's the point here... ?

The semantics from Keycloak's perspective would be - with the annotation there would be a warning or hard error, depending upon the keycloak config, when more than 1 non-XA datasource is involved in the transaction. Without the annotation, no warning or error is expected.

**Comment by shawkins** [comment]
Now that we are finalizing our upgrade to 3.13 the thinking is that it will be simipler to not fight the direction of the safety check, and instead cut off support for multi-source scenarios with non-XA datasources. This should impact only a small amount of potential keyclaok usage, and we can assess any potential workarounds later if needed.

cc @vmuzikar @ahus1

**Reply by yrodiere** [reply]
That definitely looks like the easiest path if it's possible. Thank you for the update!

**Reply by vmuzikar** [reply]
This will be addressed in:
* https://github.com/keycloak/keycloak/pull/32403
* https://github.com/keycloak/keycloak-quickstarts/pull/588
